src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs

```diff
@@ -1597,17 +1598,23 @@ private BoundNode RewriteLambdaConversion(BoundLambda node)
                         // different from the local variable `type`, which has the node's type substituted for the current container.
                         var cacheVariableType = containerAsFrame.TypeMap.SubstituteType(node.Type).Type;
 
-                        var cacheVariableName = GeneratedNames.MakeLambdaCacheFieldName(
-                            // If we are generating the field into a display class created exclusively for the lambda the lambdaOrdinal itself is unique already, 
-                            // no need to include the top-level method ordinal in the field name.
-                            (closureKind == ClosureKind.General) ? -1 : topLevelMethodId.Ordinal,
-                            topLevelMethodId.Generation,
-                            lambdaId.Ordinal,
-                            lambdaId.Generation);
-
-                        var cacheField = new SynthesizedLambdaCacheFieldSymbol(translatedLambdaContainer, cacheVariableType, cacheVariableName, _topLevelMethod, isReadOnly: false, isStatic: closureKind == ClosureKind.Singleton);
-                        CompilationState.ModuleBuilderOpt.AddSynthesizedDefinition(translatedLambdaContainer, cacheField);
-                        cache = F.Field(receiver, cacheField.AsMember(constructedFrame)); //NOTE: the field was added to the unconstructed frame type.
+                        // We cannot reference a type argument from the parent method outside the method
+                        // so if the lambda has such an argument we cannot cache it in a field
+                        if (!HasTypeArgumentsFromReferencedMethod(cacheVariableType))
```

Comment:
>HasTypeArgumentsFromReferencedMethod(cacheVariableType) [](start = 29, length = 55)

I would prefer this logic to be pulled out of the entire ```if (shouldCacheForStaticMethod || shouldCacheInLoop)``` and be incorporated into calculation of ```shouldCacheInLoop``` value.

Reply:
> I would prefer this logic to be pulled out of the entire if (shouldCacheForStaticMethod || shouldCacheInLoop) and be incorporated into calculation of shouldCacheInLoop value.

I changed my mind, let's not focus on this and keep the fix simple and local.

---
In reply to: [439131915](https://github.com/dotnet/roslyn/pull/44939#discussion_r439131915) [](ancestors = 439131915)